src/cose/cose.rs

```diff
@@ -46,8 +49,134 @@ pub fn verify_signature(payload: &[u8], cose_signature: Vec<u8>) -> Result<(), C
     Ok(())
 }
 
-/// Sign the payload and return a serialised COSE_Sign object.
-#[allow(unused_variables)]
-pub fn sign(payload: &[u8]) -> Result<Vec<u8>, CoseError> {
-    unimplemented!()
+// XXX: This works only with P256 for now!
+fn build_pkcs8(public_key: &[u8], secret_key: &[u8]) -> Vec<u8> {
+    let mut pkcs8: Vec<u8> =
+        vec![
+      0x30, 0x81, 0x87, 0x02, 0x01, 0x00, 0x30, 0x13, 0x06, 0x07, 0x2a, 0x86,
+      0x48, 0xce, 0x3d, 0x02, 0x01, 0x06, 0x08, 0x2a, 0x86, 0x48, 0xce, 0x3d,
+      0x03, 0x01, 0x07, 0x04, 0x6d, 0x30, 0x6b, 0x02, 0x01, 0x01, 0x04, 0x20,
+      // Private key: len, bytes.
+      // Public key (uncompressed EC point for now): 0xa1, len, type, len, 0x00, bytes.
+      ];
+    pkcs8.extend_from_slice(secret_key);
+    pkcs8.push(0xa1);
+    let len = (public_key.len() + 4 - 1) as u8;
+    pkcs8.push(len);
+    pkcs8.push(0x03);
+    pkcs8.push(len - 2);
+    pkcs8.push(0x00);
+    pkcs8.extend_from_slice(public_key);
+
+    return pkcs8;
+}
+
+/// RFC 8152 Section 4.4
+///    Sig_structure = [
+///        context : "Signature" / "Signature1" / "CounterSignature",
+///        body_protected : empty_or_serialized_map,
+///        ? sign_protected : empty_or_serialized_map,
+///        external_aad : bstr,
+///        payload : bstr
+///    ]
+fn build_signature_structure(payload: &[u8], alg: CoseSignatureType) -> Vec<u8> {
+    let mut sig_structure: Vec<CborType> = Vec::new();
+    sig_structure.push(CborType::String("Signature".to_string()));
+    // TODO: add the protected body header here.
```

Comment:
Yeah this seems like something we should do.

Reply:
I filed issue #15 to make the signature compliant with the way we defined add-on signatures.